aten/src/ATen/native/cuda/CuFFTPlanCache.h

```diff
@@ -109,153 +258,76 @@ class CuFFTConfig {
-        TORCH_CHECK(is_pow_of_two(signal_size),
-                 "cuFFT doesn't support signals of half type with size at any ",
-                 "dimension that is not a power of two, but got a signal size of ",
-                 checked_signal_sizes);
+        TORCH_CHECK(is_pow_of_two(sizes[i + 1]),
+            "cuFFT doesn't support signals of half type with size at any ",
+            "dimension that is not a power of two, but got a signal size of ",
+            sizes.slice(1));
       }
-      clone_input |= input.stride(signal_ndim) != 1;
-    }
-
-    // check the input sizes and strides to see if we need to make it contiguous
-    // cuFFT doesn't support batch dim with stride 0
-    clone_input |= input.stride(0) == 0;
-
-    if (complex_input) {
-      // Real/imag dimension must be like complex type.
-      clone_input |= input.stride(-1) != 1;
-      // Strides of other dimensions needs to be aligned when viewed as of complex
-      // type, i.e., multiples of 2. We check the batch dim and last signal dim
-      // here. If the input can be viewed as having embedded strides, the other
-      // signal dims will also satisfy this.
-      // See NOTE [ cuFFT Embedded Strides ] in native/cuda/SpectralOps.cu.
-      clone_input |= (batch > 0 && input.stride(0) % 2 != 0) ||
-                      input.stride(signal_ndim) % 2 != 0;
-
-      // Complex to real FFTs may overwrite the input buffer (gh-34551)
-      clone_input |= !complex_output;
+      clone_input |= in_strides.back() != 1;
     }
 
-    // Checks if input strides can be viewed as embedded.
-    // See NOTE [ cuFFT Embedded Strides ].
-    //
-    // TODO: Figure out why windows fails to compile
-    //         c10::optional<std::vector<long long int>> inembed_opt =
-    //         c10::nullopt;
-    //       Then move the following to a helper function.
-#ifdef __HIP_PLATFORM_HCC__
-    std::vector<int> inembed(signal_ndim);
-#else
-    std::vector<long long int> inembed(signal_ndim);
-#endif
-    if (!clone_input) {
-      auto istrides = input.strides();
-      auto last_istride = istrides[signal_ndim];
-      clone_input = last_istride <= 0;
-      for (auto i = signal_ndim - 1; !clone_input && i > 0 /* inembed[0] doesn't matteer */; i--) {
-        auto istride = istrides[i];
-        if (istride > 0 && istride % last_istride == 0) {
-          inembed[i] = istride / last_istride;
-          last_istride = istride;
-        } else {
-          clone_input = true;
-        }
-      }
+    CuFFTDataLayout in_layout;
+    if (clone_input) {
+      in_layout = cufft_simple_embed(sizes, fft_type == CuFFTTransformType::C2R);
+    } else {
+      in_layout = as_cufft_embed(in_strides, sizes, fft_type == CuFFTTransformType::C2R);
     }
+    auto out_layout = as_cufft_embed(out_strides, sizes, fft_type == CuFFTTransformType::R2C);
+    TORCH_CHECK(!out_layout.must_clone, "Out strides cannot be represented as CuFFT embedding");
```

Comment:
What should the user do if this occurs?

Reply:
This isn't up to the PyTorch user, so changed to `TORCH_INTERNAL_ASSERT`.